linear-regression/linear-regression/helper_functions.py

```diff
@@ -0,0 +1,380 @@
+            yval_original = target_transformer.inverse_transform(yval.reshape(-1,1)).flatten()
+
+            rmse = mean_squared_error(y_pred_test_original, yval_original)**0.5
+
+            # If the current rmse is lower than the best rmse, update the best rmse, worst feature, and best model
+            if rmse < best_rmse:
+                best_rmse = rmse
+                worst_feature = feat
+                best_model = model
+
+        # Save best model in the list
+        best_models.append(best_model)
+
+        # Updating variables for next loop
+        remaining_features.remove(worst_feature)
+
+        # Save (append) the best_rmse in the rmse_list
+        rmse_list.append(rmse)
+
+        # Save the remaining features in the features_list
+        features_list[i] = remaining_features.copy()
+
+    # Find the index of the best model based on the lowest RMSE
+    best_idx = np.argmin(rmse_list)
+
+    # Return the index of the best model, the features used, and the RMSE values
+    return best_idx, features_list, rmse_list
+
+def build_lasso_model(X_train, y_train, alpha):
+    """
+    Build a LASSO regression model with the given regularization constant (alpha).
+    This function should also split the data into train and validation subsets and
+    the RMSE should be calculated based on the validation data.
+
+    Args:
+        X_train: pandas DataFrame - Input DataFrame containing all training data.
+        y_train: pandas Series - Series containing target values for training data.
+        alpha: int - regularization constant.
+
+    Returns:
+        feature_coefs: list[float] - Coefficients of features.
+        best_rmse: float - RMSE for best model.
+    """
+
+    from sklearn.linear_model import Lasso
+    from sklearn.model_selection import train_test_split
+
+    # split the data into training and validation sets (xtrain, xval, ytrain, yval)
+    # use test size 0.2 and random state 1
+    xtrain, xval, ytrain, yval = train_test_split(
+        X_train, y_train, test_size=0.2, random_state=1
+    )
+
+    # transform the target variable using the transform_target function
+    ytrain, yval, target_transformer = transform_target(ytrain, yval)
+    
+    # Create Lasso object and fit model
+    lasso = Lasso(alpha = alpha)
+    model = lasso.fit(xtrain,ytrain)
+
+    # Set number of observations (n) and number of non-zero parameters (p)
+    n = xtrain.shape[0]
+    p = np.count_nonzero(model.coef_)
+
+    # Predict validation set and convert target back to original units
+    # Hint: Call reshape(-1, 1) on predictions before inverse transforming
+    preds = model.predict(xval).reshape(-1,1)
+    yval = target_transformer.inverse_transform(yval.reshape(-1,1)).flatten()
+    preds = target_transformer.inverse_transform(preds)
+
+    # Calculate RMSE based on number of non-zero coefficients
+    sse = sum((preds-yval)**2)
+    rmse = (sse/n)**0.5
```

Comment:
This is technically correct, but it uses the value of RMSE that doesn't account for degrees of freedom